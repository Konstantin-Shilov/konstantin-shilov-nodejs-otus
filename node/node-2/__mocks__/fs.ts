import * as path from "path";

const fs = jest.genMockFromModule('fs');

let mockFiles: object = Object.create(null);

function __setMockFiles(newMockFiles: object) {
  mockFiles = Object.create(null);
  for (const file in newMockFiles) {
    const dir = path.dirname(file);

    if (!mockFiles[dir]) {
      mockFiles[dir] = [];
    }
    mockFiles[dir].push(path.basename(file));
  }
}

function readdirSync(directoryPath: string) {
  return mockFiles[directoryPath] || [];
}

function statSync(filePath: string) {
  return {
    isDirectory: function() {
      return Object.keys(mockFiles).includes(filePath);
    }
  };
}

(fs as any).__setMockFiles = __setMockFiles;
(fs as any).readdirSync = readdirSync;
(fs as any).statSync = statSync;

module.exports = fs;
