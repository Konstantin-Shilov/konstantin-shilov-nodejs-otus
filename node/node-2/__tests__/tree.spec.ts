import * as fs from "fs";

const path = require('path');

const {ProcessingDir} = require('../modules/readDir.ts');
const {ProcessingTree} = require('../modules/createTree.ts');

jest.mock('fs');

const MOCK_FILE_INFO = {
  [path.resolve(__dirname, '../file1.js')]: '',
  [path.resolve(__dirname, '../file2.txt')]: '',

  [path.resolve(__dirname, '../dir')]: '',
  [path.resolve(__dirname, '../dir/test.js')]: '',

  [path.resolve(__dirname, '../dir2')]: '',
  [path.resolve(__dirname, '../dir2/dir3')]: '',
  [path.resolve(__dirname, '../dir2/dir3/test2.doc')]: '',

  [path.resolve(__dirname, '../dir4')]: '',
  [path.resolve(__dirname, '../dir4/test.js')]: '',

  [path.resolve(__dirname, '../dir5')]: '',
  [path.resolve(__dirname, '../dir5/dir6')]: '',
  [path.resolve(__dirname, '../dir5/dir6/test3.doc')]: '',
  [path.resolve(__dirname, '../dir5/dir6/dir7')]: '',
  [path.resolve(__dirname, '../dir5/dir6/dir7/test4.doc')]: '',
};
require('fs').__setMockFiles(MOCK_FILE_INFO);

describe('readDir module tests', () => {
  it('Object with processed tree should bedefined', () => {
    const {treeObject} = ProcessingDir.readDir(path.resolve(__dirname, '../'), 2, []);
    expect(treeObject).toBeDefined();
  });

  it('Parent dir should contains 3 or more items', () => {
    const {treeObject} = ProcessingDir.readDir(path.resolve(__dirname, '../'), 2, []);
    expect(treeObject.items.length).toBeGreaterThanOrEqual(3);
  });

  it('The number of directories and files should not be null', () => {
    const {dirStat} = ProcessingDir.readDir(path.resolve(__dirname, '../'), 0, []);
    expect(dirStat.directories).not.toBeNull();
    expect(dirStat.files).not.toBeNull();
  });
});

describe('createTree module tests', () => {
  it('Output should contain "test"', () => {
    const output = ProcessingTree.createTree(0, {name: 'test'});
    expect(output).toMatch('test');
  });
});

describe('Joint test of readDir and createTree with excluded dir', () => {
  it('Output should contain tree', () => {
    const {treeObject} = ProcessingDir.readDir(path.resolve(__dirname, '../'), 4, ['dir2']);
    const lastItem = treeObject.items.length;
    const output = ProcessingTree.createTree(lastItem, treeObject);

    expect(output).toMatchSnapshot();
  });
});
