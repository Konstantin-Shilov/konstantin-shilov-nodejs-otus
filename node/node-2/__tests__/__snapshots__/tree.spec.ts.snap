// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joint test of readDir and createTree with excluded dir Output should contain tree 1`] = `
"node-2
├── file1.js
├── file2.txt
├── dir
│    └── test.js
├── dir2
├── dir4
│    └── test.js
└── dir5
     └── dir6
         └── test3.doc
         └── dir7
             └── test4.doc
"
`;
